aten/src/ATen/native/vulkan/ops/Tensor.cpp

```diff
@@ -0,0 +1,1230 @@
+#include <ATen/native/vulkan/ops/Tensor.h>
+
+namespace at {
+namespace native {
+namespace vulkan {
+namespace ops {
+namespace {
+
+VkFormat convert(const caffe2::TypeMeta dtype) {
+  switch (c10::typeMetaToScalarType(dtype)) {
+    case kFloat:
+      // VK_FORMAT_R32G32B32A32_SFLOAT?
+      return VK_FORMAT_R16G16B16A16_SFLOAT;
+
+    default:
+      TORCH_CHECK(
+        false,
+        "Vulkan tensor format not supported!");
+  }
+
+  return VK_FORMAT_UNDEFINED;
+}
+
+vTensor::Access::Flags convert(const VkAccessFlags vk_access) {
+  vTensor::Access::Flags access = 0u;
+
+  constexpr VkAccessFlags kRead =
+      VK_ACCESS_HOST_READ_BIT |
+      VK_ACCESS_MEMORY_READ_BIT |
+      VK_ACCESS_SHADER_READ_BIT |
+      VK_ACCESS_TRANSFER_READ_BIT |
+      VK_ACCESS_UNIFORM_READ_BIT;
+
+  constexpr VkAccessFlags kWrite =
+      VK_ACCESS_HOST_WRITE_BIT |
+      VK_ACCESS_MEMORY_WRITE_BIT |
+      VK_ACCESS_SHADER_WRITE_BIT |
+      VK_ACCESS_TRANSFER_WRITE_BIT;
+
+  if (vk_access & kRead) {
+    access |= vTensor::Access::Read;
+  }
+
+  if (vk_access & kWrite) {
+    access |= vTensor::Access::Write;
+  }
+
+  return access;
+}
+
+vTensor::Buffer allocate_buffer(
+    api::Context* const context,
+    const IntArrayRef sizes,
+    const TensorOptions& options) {
+  TORCH_CHECK(!sizes.empty(), "Invalid Vulkan tensor size!");
+  verify(options);
+
+  // Forward function declaration
+  bool requires_staging(api::Context*);
+
+  const VkFlags usage = [context]() {
+    VkFlags usage = VK_BUFFER_USAGE_STORAGE_BUFFER_BIT;
+
+    if (requires_staging(context)) {
+      usage |= VK_BUFFER_USAGE_TRANSFER_SRC_BIT |
+               VK_BUFFER_USAGE_TRANSFER_DST_BIT;
+    }
+
+    return usage;
+  }();
+
+  const auto memory = [context]() -> api::Resource::Memory::Descriptor {
+    if (requires_staging(context)) {
```

Comment:
What is the semantic of requires_staging here,
From the first impression I was thinking that it means 'staging' buffer that should  be HOST_VISIBLE, but here I see that this is inverted and `if !requires_staging => HOST_VISIBLE` 
Is it right and I am missing something ?

Reply:
We will need staging buffers for discrete GPUs only.   Under those circumstances, we will have a staging buffer in system memory, a buffer in GPU memory, and an image, again, in GPU memory.  If on UMA, we can get away with only a buffer and an image, with the buffer accessible both from CPU and GPU.

What this line says is that if we require a staging buffer, i.e. we are on a discrete system, then create the buffer copy in GPU memory only (this function, allocate_buffer, only deals with the allocation of GPU-visible buffer not CPU-visible staging area), but if we do not require a staging area, meaning that we are on UMA, then make sure this GPU-visible memory / buffer is also CPU-visible, which is handled in the (implicit) 'else' part of this if statement.